jaxrs-tck/src/test/java/jakarta/ws/rs/tck/common/matchers/IsThrowingMatcherTest.java

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright (c) 2022 Jeremias Weber. All rights reserved.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0, which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the
+ * Eclipse Public License v. 2.0 are satisfied: GNU General Public License,
+ * version 2 with the GNU Classpath Exception, which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ */
+
+package jakarta.ws.rs.tck.common.matchers;
+
+import org.junit.jupiter.api.Test;
+
+import static ee.jakarta.tck.ws.rs.common.matchers.IsThrowingMatcher.isThrowing;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.not;
+
+final class IsThrowingMatcherTest {
+
+    @Test 
+    final void shouldMatchIfIsInstance() {
+        assertThat(
+                // when
+                () -> { throw new ChildException(); }, 
+                // then
+                isThrowing(ParentException.class)
+        );
+    }
+
+    @Test
+    final void shouldNotMatchIfIsNotInstance() {
```

Comment:
Typically unit tests will never include inverse tests ("should not") as the list of what it shall *not* do would be endless.

Reply:
Ok, but I think these cases are necessary to prove the matchers correctness.

Reply:
Why? A matcher only needs to guarantee to match a case, but not to guarantee to *not* match other cases. This is what I mean with "inverse".

Reply:
If I wrote a Matcher that always returned true, the first test case would succeed. So according to you the Matcher would then be correct, yes? But in fact it is not correct. That's why I think it is essential to have one examplary test for each case where it should not match as well. 

I'm not saying that we should test every possible case because as you already stated, they would be endless. But if possibilities being endless is the problem, then we could argue that the case where the matcher should match has endless possibilities as well and we cannot test all of them.

I just think this is one of the cases where a convention or best practice is not absolute and should allow for an exception.

Reply:
You misunderstood what unit tests are good for. If you implement a feature, you only need to proof that the feature *exists*, but not that it is *absolutely bug free*. The latter is inherently impossible. If your matcher matches in wrong cases, that is simply a *bug*, so *once that bug really exists* you can add a test that proofs *that* bug is gone *then*. It is impossibly to know *all* bugs upfront, so why testing *exactling this one*?

And BTW, this discussion has an absolute answer, if you ask people that *really* dominate the area of unit testing: *Here* you need a proof that a feature *exists*, and once you discover a real bug *then* you need a regression test. Anything else voids Knuth's paradigm: Premature optimization is the root of all evil. ;-)

Reply:
Ok, so I was right in my logic but wrong in my assumptions.
And TIL that I've been regression testing all along when I thought I was unit testing. 😅

It's definitely an interesting take on efficiency. Because why should I write complex tests to show that a really easy to implement feature (such as this) works? Unless it doesn't. And then I can write regression tests.

One question I still have is: But what about much more complex features?
I guess the answer is to break the complex parts down into small and easy chunks and then test those the same way?